compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -1365,16 +1432,31 @@ impl<'tcx> MutVisitor<'tcx> for VnState<'_, 'tcx> {
                 .as_local()
                 .and_then(|local| self.locals[local])
                 .or_else(|| self.simplify_rvalue(rvalue, location));
+            debug!(?value);
             let Some(value) = value else { return };
 
-            if let Some(const_) = self.try_as_constant(value) {
+            debug!(before_rvalue = ?rvalue);
+            // Ignore arrays in operands.
+            // Prevent code bloat that makes
+            // `_2 = _1` now resolved to `_2 = <evaluated array>`.
+            let disallow_dup_array = if rvalue.ty(self.local_decls, self.tcx).is_array()
+                && let Some(locals) = self.rev_locals.get(value).as_deref()
+                && let [first, ..] = locals[..]
+            {
+                first != lhs.local
+            } else {
+                false
+            };
```

Comment:
I don't understand what this is meant to do. Why is this only checking the first inside `rev_local`?

Reply:
Without it, duplicate arrays in code will be duplicated in mir output.
Consider this
```rust
// code
a = [1,2,3];
b = [1,2,3];
// mir without this condition.
_1 = const [1,2,3];
_2 = const [1,2,3];
// mir with this condition
_1 = const [1,2,3];
_2 = _1;
```